src/Http/Routing/src/Matching/HttpMethodMatcherPolicy.cs

```diff
@@ -62,6 +83,104 @@ public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
             return false;
         }
 
+        /// <summary>
+        /// For framework use only.
+        /// </summary>
+        /// <param name="httpContext"></param>
+        /// <param name="context"></param>
+        /// <param name="candidates"></param>
+        /// <returns></returns>
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            // Returning a 405 here requires us to return keep track of all 'seen' HTTP methods. We allocate to
+            // keep track of this beause we either need to keep track of the HTTP methods or keep track of the
+            // endpoints - both allocate.
+            //
+            // Those code only runs in the presence of dynamic endpoints anyway.
+            //
+            // We want to return a 405 iff we eliminated ALL of the currently valid endpoints due to HTTP method
+            // mismatch.
+            bool? needs405Endpoint = null;
+            HashSet<string> methods = null;
+
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                // We do this check first for consistency with how 405 is implemented for the graph version
+                // of this code. We still want to know if any endpoints in this set require an HTTP method
+                // even if those endpoints are already invalid.
+                var metadata = candidates[i].Endpoint.Metadata.GetMetadata<IHttpMethodMetadata>();
+                if (metadata == null || metadata.HttpMethods.Count == 0)
+                {
+                    // Can match any method.
+                    needs405Endpoint = false;
+                    continue;
+                }
+
+                // Saw a valid endpoint.
+                needs405Endpoint = needs405Endpoint ?? true;
+
+                if (!candidates.IsValidCandidate(i))
+                {
+                    continue;
+                }
+
+                var httpMethod = httpContext.Request.Method;
+                if (metadata.AcceptCorsPreflight &&
+                    string.Equals(httpMethod, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
+                    httpContext.Request.Headers.ContainsKey(OriginHeader) &&
+                    httpContext.Request.Headers.TryGetValue(AccessControlRequestMethod, out var accessControlRequestMethod) &&
```

Comment:
`Request.Headers.TryGetValue(AccessControlRequestMethod` 
=> `Request.Headers.TryGetValue(HeaderNames.AccessControlRequestMethod`

Reply:
Also lift Request.Headers to a local rather than looking it up via 2 interfaces twice
```csharp
var headers = httpContext.Request.Headers;
if (metadata.AcceptCorsPreflight &&
    string.Equals(httpMethod, PreflightHttpMethod, StringComparison.OrdinalIgnoreCase) &&
    headers.ContainsKey(HeaderNames.OriginHeader) &&
    headers.TryGetValue(HeaderNames.AccessControlRequestMethod, out var accessControlRequestMethod) &&
    !StringValues.IsNullOrEmpty(accessControlRequestMethod))
```

e.g. non-faded are calls made getting item from RequestHeaders via httpContext:

![image](https://user-images.githubusercontent.com/1142958/56105979-d8b31b00-5f36-11e9-8dcb-89ada2e34471.png)


The long time spent in string.Equals, string.EqualsOridinalIgnoreCase for a 4 letter string "Host" is what lead to #9341 and its acceleration for `HeaderNames.` strings

Reply:
Much <3 thanks!